Mage.Sets/src/mage/sets/masterseditionii/SoulExchange.java

```diff
@@ -0,0 +1,111 @@
+ *  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ *  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *  The views and conclusions contained in the software and documentation are those of the
+ *  authors and should not be interpreted as representing official policies, either expressed
+ *  or implied, of BetaSteward_at_googlemail.com.
+ */
+package mage.sets.masterseditionii;
+
+import java.util.UUID;
+import mage.abilities.Ability;
+import mage.abilities.costs.Cost;
+import mage.abilities.costs.common.ExileTargetCost;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.ReturnFromGraveyardToBattlefieldTargetEffect;
+import mage.cards.CardImpl;
+import mage.constants.CardType;
+import mage.constants.Outcome;
+import mage.constants.Rarity;
+import mage.counters.CounterType;
+import mage.filter.common.FilterCreatureCard;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.target.common.TargetCardInYourGraveyard;
+import mage.target.common.TargetControlledCreaturePermanent;
+
+/**
+ *
+ * @author MarcoMarin
+ */
+public class SoulExchange extends CardImpl {
+
+    public SoulExchange(UUID ownerId) {
+        super(ownerId, 111, "Soul Exchange", Rarity.UNCOMMON, new CardType[]{CardType.SORCERY}, "{B}{B}");
+        this.expansionSetCode = "ME2";
+
+        // As an additional cost to cast Soul Exchange, exile a creature you control.
+        Cost cost = new ExileTargetCost(new TargetControlledCreaturePermanent());
+        this.getSpellAbility().addCost(cost);
+        // Return target creature card from your graveyard to the battlefield. Put a +2/+2 counter on that creature if the exiled creature was a Thrull.
+        this.getSpellAbility().addTarget(new TargetCardInYourGraveyard(new FilterCreatureCard("creature card from your graveyard")));        
+        this.getSpellAbility().addEffect(new SoulExchangeEffect());        
+        
+    }
+
+    public SoulExchange(final SoulExchange card) {
+        super(card);
+    }
+
+    @Override
+    public SoulExchange copy() {
+        return new SoulExchange(this);
+    }
+}
+
+class SoulExchangeEffect extends OneShotEffect{
+                          
+    public SoulExchangeEffect() {
+        super(Outcome.Benefit);
+        this.setText("Return target creature card from your graveyard to the battlefield. Put a +2/+2 counter on that creature if the exiled creature was a Thrull.");
+    }
+    
+    public SoulExchangeEffect(final SoulExchangeEffect effect) {
+        super(effect);
+    }
+    
+    @Override
+    public SoulExchangeEffect copy() {
+        return new SoulExchangeEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        ReturnFromGraveyardToBattlefieldTargetEffect effect = new ReturnFromGraveyardToBattlefieldTargetEffect();
+        if (!effect.apply(game, source)) return false;
+        
+        for (Cost c : source.getCosts()){
```

Comment:
If you use game.getPermanentOrLKIBattlefield(t), you get a permanent that is on the battlefield with the sourceId of the card. So if the card was put onto the battlefield after the sacrifice, you get the new permanent on the battlefield. The sourceId of a card never changes, that's tru for all zones the object goes through. So it's sometimes important to check the zoneChangeCounter (e.g. by using MageObjectReference or directly) of the object.